File: mac/Sources/FineClipboard/Services/SyncEngine.swift
import Foundation

/// Drives end-to-end-encrypted, incremental sync against finepaste-server. Only text and
/// images sync — passwords are never sent. Each item is encrypted with the passphrase-derived
/// key before upload; the server stores ciphertext only. Mirrors the Windows SyncEngine.
///
/// Invoked from the main thread; store access is hopped onto the main actor around the
/// network calls (the store is main-thread-only).
final class SyncEngine {
    static let defaultBaseURL = "https://cassiangroup.uk/finepaste"
    private static let keychainAccount = "sync-key"
    private static let checkToken = "FINECLIP-SYNC-OK"

    private let store: Store
    private let client = SyncClient()
    private var crypto: SyncCrypto?

    init(_ store: Store) {
        self.store = store
        client.baseURL = (store.setting(Store.syncBaseUrlKey) ?? Self.defaultBaseURL).trimmedSlash()
        client.token = store.setting(Store.syncTokenKey).flatMap { $0.isEmpty ? nil : $0 }
        if let keyData = Keychain.getData(account: Self.keychainAccount) {
            crypto = SyncCrypto(key: keyData)
        }
    }

    var loggedIn: Bool { !(client.token ?? "").isEmpty }
    var hasPassphrase: Bool { crypto != nil }
    var enabled: Bool { store.setting(Store.syncEnabledKey) == "1" }
    var ready: Bool { enabled && loggedIn && crypto != nil }
    var email: String? { store.setting(Store.syncEmailKey) }
    var baseURL: String { client.baseURL }

    func setBaseURL(_ url: String) {
        client.baseURL = url.trimmingCharacters(in: .whitespaces).trimmedSlash()
        store.setSetting(Store.syncBaseUrlKey, client.baseURL)
    }

    func register(email: String, password: String) async throws {
        let r = try await client.register(email: email.trimmingCharacters(in: .whitespaces), password: password)
        saveAuth(email, r)
    }

    func login(email: String, password: String) async throws {
        let r = try await client.login(email: email.trimmingCharacters(in: .whitespaces), password: password)
        saveAuth(email, r)
    }

    private func saveAuth(_ email: String, _ r: SyncClient.AuthResult) {
        client.token = r.token
        store.setSetting(Store.syncEmailKey, email.trimmingCharacters(in: .whitespaces))
        store.setSetting(Store.syncTokenKey, r.token)
    }

    func logout() {
        client.token = nil
        store.setSetting(Store.syncTokenKey, "")
        store.setSetting(Store.syncEnabledKey, "0")
    }

    /// Redeems a VIP activation key; returns true if VIP is now active.
    func redeem(key: String) async throws -> Bool {
        let r = try await client.redeem(key: key.trimmingCharacters(in: .whitespaces))
        return r.vip
    }

    /// Sets the sync passphrase: derives + stores the key (Keychain) and a local verifier.
    func setPassphrase(_ passphrase: String) throws {
        guard let email, !email.isEmpty else {
            throw SyncClient.SyncError(message: "请先登录再设置同步口令")
        }
        let keyData = SyncCrypto.deriveKey(passphrase: passphrase, email: email)
        Keychain.setData(account: Self.keychainAccount, keyData)
        let c = SyncCrypto(key: keyData)
        crypto = c
        store.setSetting(Store.syncCheckKey, c.encryptToBase64(Data(Self.checkToken.utf8)))
    }

    func enable(_ on: Bool) {
        store.setSetting(Store.syncEnabledKey, on ? "1" : "0")
        if on { store.markAllDirty() }
    }

    /// Pushes local changes then pulls remote changes. Returns a status line.
    @discardableResult
    func syncNow() async throws -> String {
        guard ready, let crypto else {
            throw SyncClient.SyncError(message: "同步未就绪(需登录、设置口令并开启同步)")
        }
        let store = self.store

        // ---- build push payload on the main thread ----
        let prep: (items: [SyncClient.PushItem], dirtyIds: [Int64], tombUuids: [String]) =
            await MainActor.run {
                var items: [SyncClient.PushItem] = []
                let dirty = store.dirtyForSync(limit: 200)
                for it in dirty {
                    let uuid = it.syncUuid ?? UUID().uuidString.replacingOccurrences(of: "-", with: "")
                    if it.syncUuid == nil { store.assignSyncUuid(it.id, uuid) }
                    items.append(SyncClient.PushItem(
                        uuid: uuid, kind: it.kind == .image ? 1 : 0,
                        updated_at: it.updatedMs, deleted: false,
                        cipher: crypto.encryptToBase64(Self.buildEnvelope(it))))
                }
                let tombs = store.tombstones()
                for t in tombs {
                    items.append(SyncClient.PushItem(uuid: t.uuid, kind: 0, updated_at: t.updatedMs, deleted: true, cipher: ""))
                }
                return (items, dirty.map { $0.id }, tombs.map { $0.uuid })
            }

        let uploaded = prep.items.count
        if uploaded > 0 {
            _ = try await client.push(retentionDays: await MainActor.run { self.retentionDays() }, items: prep.items)
            await MainActor.run {
                for id in prep.dirtyIds { store.markSynced(id) }
                for u in prep.tombUuids { store.removeTombstone(u) }
            }
        }

        // ---- pull remote changes ----
        var cursor = await MainActor.run { Int64(store.setting(Store.syncCursorKey) ?? "0") ?? 0 }
        var downloaded = 0
        while true {
            let res = try await client.changes(since: cursor, limit: 200)

            // Resolve each change (decrypt text inline / fetch+decrypt image blobs) off the main thread.
            var resolved: [(SyncClient.Change, Envelope?)] = []
            for ch in res.changes {
                if ch.deleted { resolved.append((ch, nil)); continue }
                let cipherData: Data? = ch.kind == 1
                    ? try await client.blob(uuid: ch.uuid)
                    : ch.cipher.flatMap { Data(base64Encoded: $0) }
                guard let cipherData,
                      let plain = crypto.decryptFromBase64(cipherData.base64EncodedString()),
                      let env = try? JSONDecoder().decode(Envelope.self, from: plain) else { continue }
                resolved.append((ch, env))
            }

            let newCursor = res.cursor
            let toApply = resolved
            await MainActor.run {
                for (ch, env) in toApply {
                    if ch.deleted { store.deleteBySyncUuid(ch.uuid); continue }
                    guard let env else { continue }
                    store.upsertFromSync(
                        kind: ch.kind == 1 ? .image : .text,
                        text: env.text, data: env.img.flatMap { Data(base64Encoded: $0) },
                        preview: env.preview ?? "", html: env.html, rtf: env.rtf, ocr: env.ocr,
                        uuid: ch.uuid, updatedMs: ch.updated_at)
                }
                store.setSetting(Store.syncCursorKey, String(newCursor))
            }
            cursor = newCursor
            downloaded += res.changes.count
            if !res.has_more { break }
        }

        return "已同步 · 上传 \(uploaded) · 下载 \(downloaded)"
    }

    private func retentionDays() -> Int {
        let days = Int(store.setting(Store.expiryDaysKey) ?? "0") ?? 0
        return days <= 0 ? 183 : days // server clamps to ~6 months anyway
    }

    // Envelope JSON keys must match the Windows client exactly.
    private struct Envelope: Codable {
        var text: String?
        var html: String?
        var rtf: String?
        var preview: String?
        var ocr: String?
        var img: String?
    }

    private static func buildEnvelope(_ it: ClipItem) -> Data {
        let env = Envelope(text: it.text, html: it.html, rtf: it.rtf,
                           preview: it.preview, ocr: it.ocrText,
                           img: it.data.map { $0.base64EncodedString() })
        return (try? JSONEncoder().encode(env)) ?? Data("{}".utf8)
    }
}

private extension String {
    func trimmedSlash() -> String {
        var s = self
        while s.hasSuffix("/") { s.removeLast() }
        return s
    }
}

File: src/FineClipboard/Services/SyncEngine.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using FineClipboard.Models;

namespace FineClipboard.Services;

/// <summary>
/// Drives end-to-end-encrypted, incremental sync against finepaste-server. Only text and
/// images sync — passwords are never sent (the vault is excluded entirely). Each item is
/// encrypted with the passphrase-derived key before upload; the server stores ciphertext only.
/// </summary>
public sealed class SyncEngine
{
    public const string DefaultBaseUrl = "https://cassiangroup.uk/finepaste";
    private const string CheckToken = "FINECLIP-SYNC-OK";

    private readonly HistoryStore _store;
    private readonly SyncClient _client = new();
    private SyncCrypto? _crypto;

    public SyncEngine(HistoryStore store)
    {
        _store = store;
        _client.BaseUrl = (_store.GetSetting(HistoryStore.SyncBaseUrlKey) ?? DefaultBaseUrl).TrimEnd('/');
        _client.Token = _store.GetSetting(HistoryStore.SyncTokenKey);
        byte[]? key = UnsealKey();
        if (key != null)
        {
            _crypto = SyncCrypto.FromKey(key);
        }
    }

    public bool LoggedIn => !string.IsNullOrEmpty(_client.Token);
    public bool HasPassphrase => _crypto != null;
    public bool Enabled => _store.GetSetting(HistoryStore.SyncEnabledKey) == "1";
    public bool Ready => Enabled && LoggedIn && _crypto != null;
    public string? Email => _store.GetSetting(HistoryStore.SyncEmailKey);
    public string BaseUrl => _client.BaseUrl;

    public void SetBaseUrl(string url)
    {
        _client.BaseUrl = url.Trim().TrimEnd('/');
        _store.SetSetting(HistoryStore.SyncBaseUrlKey, _client.BaseUrl);
    }

    public async Task RegisterAsync(string email, string password)
    {
        SyncClient.AuthResult? r = await _client.RegisterAsync(email.Trim(), password);
        SaveAuth(email, r);
    }

    public async Task LoginAsync(string email, string password)
    {
        SyncClient.AuthResult? r = await _client.LoginAsync(email.Trim(), password);
        SaveAuth(email, r);
    }

    private void SaveAuth(string email, SyncClient.AuthResult? r)
    {
        if (r == null || string.IsNullOrEmpty(r.Token))
        {
            throw new SyncException("登录失败");
        }
        _client.Token = r.Token;
        _store.SetSetting(HistoryStore.SyncEmailKey, email.Trim());
        _store.SetSetting(HistoryStore.SyncTokenKey, r.Token);
    }

    public void Logout()
    {
        _client.Token = null;
        _store.SetSetting(HistoryStore.SyncTokenKey, string.Empty);
        _store.SetSetting(HistoryStore.SyncEnabledKey, "0");
    }

    /// <summary>Redeems a VIP activation key; returns true if VIP is now active.</summary>
    public async Task<bool> RedeemAsync(string key)
    {
        SyncClient.AuthResult? r = await _client.RedeemAsync(key.Trim());
        return r?.Vip ?? false;
    }

    /// <summary>Sets the sync passphrase: derives + seals the key and stores a local verifier.</summary>
    public void SetPassphrase(string passphrase)
    {
        if (string.IsNullOrEmpty(Email))
        {
            throw new SyncException("请先登录再设置同步口令");
        }
        byte[] key = SyncCrypto.DeriveKey(passphrase, Email);
        SealKey(key);
        _crypto = SyncCrypto.FromKey(key);
        _store.SetSetting(HistoryStore.SyncCheckKey,
            _crypto.EncryptToBase64(Encoding.UTF8.GetBytes(CheckToken)));
    }

    public void Enable(bool on)
    {
        _store.SetSetting(HistoryStore.SyncEnabledKey, on ? "1" : "0");
        if (on)
        {
            _store.MarkAllDirty(); // first enable: push existing text/image history
        }
    }

    /// <summary>Pushes local changes then pulls remote changes. Returns a status line.</summary>
    public async Task<string> SyncNowAsync()
    {
        if (!Ready)
        {
            throw new SyncException("同步未就绪(需登录、设置口令并开启同步)");
        }

        // ---- push dirty items + tombstones ----
        List<ClipboardItem> dirty = _store.GetDirtyForSync(200);
        var items = new List<SyncClient.PushItem>();
        foreach (ClipboardItem it in dirty)
        {
            string uuid = it.SyncUuid ?? Guid.NewGuid().ToString("N");
            if (it.SyncUuid == null)
            {
                _store.AssignSyncUuid(it.Id, uuid);
            }
            byte[] envelope = BuildEnvelope(it);
            items.Add(new SyncClient.PushItem
            {
                Uuid = uuid,
                Kind = it.Type == ClipItemType.Image ? 1 : 0,
                UpdatedAt = it.UpdatedMs,
                Deleted = false,
                Cipher = _crypto!.EncryptToBase64(envelope),
            });
        }
        List<(string Uuid, long UpdatedMs)> tombs = _store.GetTombstones();
        foreach ((string uuid, long ms) in tombs)
        {
            items.Add(new SyncClient.PushItem { Uuid = uuid, Kind = 0, UpdatedAt = ms, Deleted = true, Cipher = string.Empty });
        }

        int uploaded = items.Count;
        if (uploaded > 0)
        {
            await _client.PushAsync(RetentionDays(), items);
            foreach (ClipboardItem it in dirty)
            {
                _store.MarkSynced(it.Id);
            }
            foreach ((string uuid, long _) in tombs)
            {
                _store.RemoveTombstone(uuid);
            }
        }

        // ---- pull remote changes ----
        long cursor = long.TryParse(_store.GetSetting(HistoryStore.SyncCursorKey), out long c) ? c : 0;
        int downloaded = 0;
        while (true)
        {
            SyncClient.ChangesResult res = await _client.ChangesAsync(cursor, 200);
            foreach (SyncClient.Change ch in res.Changes)
            {
                await ApplyChangeAsync(ch);
            }
            cursor = res.Cursor;
            _store.SetSetting(HistoryStore.SyncCursorKey, cursor.ToString(System.Globalization.CultureInfo.InvariantCulture));
            downloaded += res.Changes.Count;
            if (!res.HasMore)
            {
                break;
            }
        }

        return $"已同步 · 上传 {uploaded} · 下载 {downloaded}";
    }

    private async Task ApplyChangeAsync(SyncClient.Change ch)
    {
        if (ch.Deleted)
        {
            _store.DeleteBySyncUuid(ch.Uuid);
            return;
        }

        byte[]? cipher;
        if (ch.Kind == 1)
        {
            cipher = await _client.BlobAsync(ch.Uuid); // images stored as blobs, fetched lazily
        }
        else
        {
            cipher = string.IsNullOrEmpty(ch.Cipher) ? null : Convert.FromBase64String(ch.Cipher);
        }
        if (cipher == null)
        {
            return;
        }

        byte[]? plain = _crypto!.DecryptFromBase64(Convert.ToBase64String(cipher));
        if (plain == null)
        {
            return; // wrong passphrase / corrupt — skip
        }
        ClipboardItem? item = ParseEnvelope(plain, ch.Kind);
        if (item != null)
        {
            _store.UpsertFromSync(item, ch.Uuid, ch.UpdatedAt);
        }
    }

    private int RetentionDays()
    {
        int days = int.TryParse(_store.GetSetting(HistoryStore.ExpiryDaysKey), out int d) ? d : 0;
        return days <= 0 ? 183 : days; // server clamps to ~6 months anyway
    }

    private sealed class Envelope
    {
        [JsonPropertyName("text")] public string? Text { get; set; }
        [JsonPropertyName("html")] public string? Html { get; set; }
        [JsonPropertyName("rtf")] public string? Rtf { get; set; }
        [JsonPropertyName("preview")] public string? Preview { get; set; }
        [JsonPropertyName("ocr")] public string? Ocr { get; set; }
        [JsonPropertyName("img")] public string? Img { get; set; }
    }

    private static byte[] BuildEnvelope(ClipboardItem it)
    {
        var env = new Envelope
        {
            Text = it.Text,
            Html = it.Html,
            Rtf = it.Rtf,
            Preview = it.Preview,
            Ocr = it.OcrText,
            Img = it.ImageData != null ? Convert.ToBase64String(it.ImageData) : null,
        };
        return JsonSerializer.SerializeToUtf8Bytes(env);
    }

    private static ClipboardItem? ParseEnvelope(byte[] plain, int kind)
    {
        Envelope? env;
        try
        {
            env = JsonSerializer.Deserialize<Envelope>(plain);
        }
        catch
        {
            return null;
        }
        if (env == null)
        {
            return null;
        }
        return new ClipboardItem
        {
            Type = kind == 1 ? ClipItemType.Image : ClipItemType.Text,
            Text = env.Text,
            Html = env.Html,
            Rtf = env.Rtf,
            Preview = env.Preview ?? string.Empty,
            OcrText = env.Ocr,
            ImageData = env.Img != null ? Convert.FromBase64String(env.Img) : null,
        };
    }

    private void SealKey(byte[] key)
    {
        byte[] sealedKey = ProtectedData.Protect(key, null, DataProtectionScope.CurrentUser);
        _store.SetSetting(HistoryStore.SyncKeyKey, Convert.ToBase64String(sealedKey));
    }

    private byte[]? UnsealKey()
    {
        string? s = _store.GetSetting(HistoryStore.SyncKeyKey);
        if (string.IsNullOrEmpty(s))
        {
            return null;
        }
        try
        {
            return ProtectedData.Unprotect(Convert.FromBase64String(s), null, DataProtectionScope.CurrentUser);
        }
        catch
        {
            return null;
        }
    }
}
